rust/module/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+                    fn read(&self) -> {param_type} {{
+                        unsafe {{ __{name}_{param_name}_VALUE }}
+                    }}
+                }}
+
+                const {param_name}: __{name}_{param_name} = __{name}_{param_name};
+
+                // FIXME: does the `align` do the right thing here?
+                // `core::mem::size_of(usize)`
+                #[repr(C,align(8))]
+                struct __{name}_{param_name}_RacyKernelParam(kernel::bindings::kernel_param);
+
+                unsafe impl Sync for __{name}_{param_name}_RacyKernelParam {{
+                }}
+
+                #[cfg(not(MODULE))]
+                const __{name}_{param_name}_NAME: *const kernel::c_types::c_char = b\"{name}.{param_name}\\0\" as *const _ as *const kernel::c_types::c_char;
+
+                #[cfg(MODULE)]
+                const __{name}_{param_name}_NAME: *const kernel::c_types::c_char = b\"{param_name}\\0\" as *const _ as *const kernel::c_types::c_char;
+
+                #[link_section = \"__param\"]
+                #[used]
+                static __{name}_{param_name}_STRUCT: __{name}_{param_name}_RacyKernelParam = __{name}_{param_name}_RacyKernelParam(kernel::bindings::kernel_param {{
+                    name: __{name}_{param_name}_NAME,
+                    // TODO: `THIS_MODULE`
+                    mod_: core::ptr::null_mut(),
+                    ops: unsafe {{ &kernel::bindings::param_ops_{param_kernel_type} }} as *const kernel::bindings::kernel_param_ops,
+                    perm: {permissions},
+                    level: -1,
+                    flags: 0,
+                    __bindgen_anon_1: kernel::bindings::kernel_param__bindgen_ty_1 {{
+                        arg: unsafe {{ &__{name}_{param_name}_VALUE }} as *const _ as *mut kernel::c_types::c_void,
+                    }},
+                }});
+                ",
+                name = name,
+                param_type = param_type,
+                param_kernel_type = param_kernel_type,
+                param_default = param_default,
+                param_name = param_name,
+                permissions = param_permissions,
+            )
+        );
+    }
+
+    format!(
+        "
+            static mut __MOD: Option<{typename}> = None;
+
+            // Loadable modules need to export the `{{init,cleanup}}_module` identifiers
+            #[cfg(MODULE)]
+            #[no_mangle]
+            pub extern \"C\" fn init_module() -> kernel::c_types::c_int {{
+                __init()
+            }}
+
+            #[cfg(MODULE)]
+            #[no_mangle]
+            pub extern \"C\" fn cleanup_module() {{
+                __exit()
+            }}
+
+            // Built-in modules are initialized through an initcall pointer
+            // and the identifiers need to be unique
+            #[cfg(not(MODULE))]
+            #[cfg(not(CONFIG_HAVE_ARCH_PREL32_RELOCATIONS))]
+            #[link_section = \".initcall6.init\"]
```

Comment:
I could use a constant for the link section since it is used twice, which also helps if it changes in the future.